// @arcana/ca-common@1.0.1-alp/dist/cjs/vaultcontracts/vaultcontracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Environment = void 0;
exports.getVaultContractMap = getVaultContractMap;
const viem_1 = require("viem");
const data_1 = require("../data");
const definition_1 = require("../proto/definition");
var Environment;
(function (Environment) {
    Environment[Environment["FOLLY"] = 0] = "FOLLY";
    Environment[Environment["CERISE"] = 1] = "CERISE";
    Environment[Environment["CORAL"] = 2] = "CORAL";
    Environment[Environment["JADE"] = 3] = "JADE";
})(Environment || (exports.Environment = Environment = {}));
const dataSets = new Map([
    [
        Environment.FOLLY,
        [
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 421614),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 11155420),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 80002),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 84532),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 11155111),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 10143),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
        ],
    ],
    [
        Environment.CERISE,
        [
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 137),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 10),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 42161),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 8453),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 43114),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 534352),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 8217),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 50104),
                "0x0A8eb0237524D1A8Fa8cbecF49e54FE627Ed781f",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 56),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 999),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.FUEL, 9889),
                "0x6cd9b8d7e13762f4cb98cbd733640138aeb65395c4b38d84a7d46f6f6c6b42e4",
            ],
        ],
    ],
    [
        Environment.CORAL,
        [
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 1),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 10),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 137),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 42161),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 534352),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 8453),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 43114),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 999),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 8217),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 50104),
                "0xB61fAdeBccCb15823b64bf47829d32eeb4A08930",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.ETHEREUM, 56),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new data_1.OmniversalChainID(definition_1.Universe.FUEL, 9889),
                "0xe2586f908cc885e630cec6d8d578f02e6ade66983baf23f82757be502127dfb1",
            ],
        ],
    ],
]);
function getVaultContractMap(environment) {
    const src = dataSets.get(environment);
    if (src == null) {
        throw new Error("Environment not found");
    }
    return new data_1.ChainIDKeyedMap(src.map((t) => [
        t[0],
        (0, data_1.zeroExtendBufToGivenSize)((0, viem_1.hexToBytes)(t[1]), 32),
    ]));
}
