// @arcana/ca-common@1.0.1-alp/dist/esm/vaultcontracts/vaultcontracts.js
import { hexToBytes } from "viem";
import { ChainIDKeyedMap, OmniversalChainID, zeroExtendBufToGivenSize, } from "../data";
import { Universe } from "../proto/definition";
export var Environment;
(function (Environment) {
    Environment[Environment["FOLLY"] = 0] = "FOLLY";
    Environment[Environment["CERISE"] = 1] = "CERISE";
    Environment[Environment["CORAL"] = 2] = "CORAL";
    Environment[Environment["JADE"] = 3] = "JADE";
})(Environment || (Environment = {}));
const dataSets = new Map([
    [
        Environment.FOLLY,
        [
            [
                new OmniversalChainID(Universe.ETHEREUM, 421614),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 11155420),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 80002),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 84532),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 11155111),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 10143),
                "0xF0111EdE031a4377C34A4AD900f1E633E41055Dc",
            ],
        ],
    ],
    [
        Environment.CERISE,
        [
            [
                new OmniversalChainID(Universe.ETHEREUM, 137),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 10),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 42161),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 8453),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 43114),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 534352),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 8217),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 50104),
                "0x0A8eb0237524D1A8Fa8cbecF49e54FE627Ed781f",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 56),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 999),
                "0x0000002Ed0657b924b4AA83aD76CaB42DF90869D",
            ],
            [
                new OmniversalChainID(Universe.FUEL, 9889),
                "0x6cd9b8d7e13762f4cb98cbd733640138aeb65395c4b38d84a7d46f6f6c6b42e4",
            ],
        ],
    ],
    [
        Environment.CORAL,
        [
            [
                new OmniversalChainID(Universe.ETHEREUM, 1),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 10),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 137),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 42161),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 534352),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 8453),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 43114),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 999),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 8217),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 50104),
                "0xB61fAdeBccCb15823b64bf47829d32eeb4A08930",
            ],
            [
                new OmniversalChainID(Universe.ETHEREUM, 56),
                "0xBADA557252D286e45a1AD73f32479062D4E2e86B",
            ],
            [
                new OmniversalChainID(Universe.FUEL, 9889),
                "0xe2586f908cc885e630cec6d8d578f02e6ade66983baf23f82757be502127dfb1",
            ],
        ],
    ],
]);
export function getVaultContractMap(environment) {
    const src = dataSets.get(environment);
    if (src == null) {
        throw new Error("Environment not found");
    }
    return new ChainIDKeyedMap(src.map((t) => [
        t[0],
        zeroExtendBufToGivenSize(hexToBytes(t[1]), 32),
    ]));
}
